concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Concordium.KonsensusV1.Types where
+
+import Control.Monad
+import Data.Bits
+import qualified Data.ByteString as BS
+import qualified Data.Map.Strict as Map
+import Data.Serialize
+import qualified Data.Vector as Vector
+import Data.Void
+import Data.Word
+import Numeric.Natural
+
+import qualified Concordium.Crypto.BlockSignature as BlockSig
+import qualified Concordium.Crypto.BlsSignature as Bls
+import qualified Concordium.Crypto.SHA256 as Hash
+import qualified Concordium.Crypto.VRF as VRF
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.Transactions
+import Concordium.Utils.Serialization
+
+import qualified Concordium.GlobalState.Basic.BlockState.LFMBTree as LFMBT
+
+-- |A round number for consensus.
+newtype Round = Round {theRound :: Word64}
+    deriving (Eq, Ord, Show, Serialize, Num, Integral, Real, Enum, Bounded)
+
+-- |A strict version of 'Maybe'. We deliberately avoid defining generalised serialization and
+-- hashing instances, so that specific instances can be given as appropriate.
+data Optionally a
+    = Absent
+    | Present !a
+    deriving (Eq, Ord, Show, Functor, Foldable, Traversable)
+
+-- |Returns 'True' if and only if the value is 'Present'.
+isPresent :: Optionally a -> Bool
+isPresent Absent = False
+isPresent (Present _) = True
+
+-- |Returns 'True' if and only if the value is 'Absent'.
+isAbsent :: Optionally a -> Bool
+isAbsent Absent = True
+isAbsent (Present _) = False
+
+-- |The message that is signed by a finalizer to certify a block.
+data QuorumSignatureMessage = QuorumSignatureMessage
+    { -- |Hash of the genesis block.
+      qsmGenesis :: !BlockHash,
+      -- |Hash of the block being signed.
+      qsmBlock :: !BlockHash
+    }
+    deriving (Eq, Show)
+
+-- |Compute the byte representation of a 'QuorumSignatureMessage' that is actually signed.
+-- TODO: check that this cannot collide with other signed messages.
+quorumSignatureMessageBytes :: QuorumSignatureMessage -> BS.ByteString
+quorumSignatureMessageBytes QuorumSignatureMessage{..} = runPut $ do
+    putByteString "QUORUM."
+    put qsmGenesis
+    put qsmBlock
+
+-- |Signature by a finalizer on a 'QuorumSignatureMessage', or an aggregation of signatures on a
+-- common message.
+newtype QuorumSignature = QuorumSignature {theQuorumSignature :: Bls.Signature}
+    deriving (Eq, Ord, Show, Serialize, Semigroup, Monoid)
```

Comment:
I am really not a fan of this being a "Monoid". The unit for this is a signature that really should never be accepted, but accidentally could be if care is not taken.